File: java-springboot/src/main/java/com/benchmark/orders/application/CreateOrderUseCase.java
package com.benchmark.orders.application;

import com.benchmark.orders.application.dto.CreateOrderInput;
import com.benchmark.orders.application.dto.CreateOrderItemInput;
import com.benchmark.orders.application.dto.CreateOrderOutput;
import com.benchmark.orders.domain.Order;
import com.benchmark.orders.domain.OrderItem;
import com.benchmark.orders.domain.OrderRepository;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
public class CreateOrderUseCase {

    private final OrderRepository repo;

    public CreateOrderUseCase(OrderRepository repo) {
        this.repo = repo;
    }

    public CreateOrderOutput execute(CreateOrderInput input) {
        if (input.customerId() == null || input.customerId().isBlank()) {
            throw new IllegalArgumentException("customerId is required");
        }
        if (input.items() == null || input.items().isEmpty()) {
            throw new IllegalArgumentException("items must not be empty");
        }

        List<OrderItem> items = new ArrayList<>();
        for (int i = 0; i < input.items().size(); i++) {
            CreateOrderItemInput item = input.items().get(i);
            if (item.productId() == null || item.productId().isBlank()) {
                throw new IllegalArgumentException("item[" + i + "].productId is required");
            }
            if (item.quantity() <= 0) {
                throw new IllegalArgumentException("item[" + i + "].quantity must be greater than 0");
            }
            if (item.price() < 0) {
                throw new IllegalArgumentException("item[" + i + "].price must be >= 0");
            }
            items.add(new OrderItem(item.productId(), item.quantity(), item.price()));
        }

        Order order = Order.create(input.customerId(), items);
        repo.save(order);

        return new CreateOrderOutput(
                order.getOrderId(),
                order.getCustomerId(),
                order.getTotalAmount(),
                order.getItemsCount(),
                order.getCreatedAt()
        );
    }
}

File: go-stdlib/internal/application/create_order_usecase.go
package application

import (
	"context"
	"fmt"

	"github.com/benchmark/orders/internal/domain"
)

type CreateOrderUseCase struct {
	repo domain.OrderRepository
}

func NewCreateOrderUseCase(repo domain.OrderRepository) *CreateOrderUseCase {
	return &CreateOrderUseCase{repo: repo}
}

func (uc *CreateOrderUseCase) Execute(ctx context.Context, input CreateOrderInput) (*CreateOrderOutput, error) {
	if input.CustomerID == "" {
		return nil, fmt.Errorf("customerId is required")
	}
	if len(input.Items) == 0 {
		return nil, fmt.Errorf("items must not be empty")
	}

	items := make([]domain.OrderItem, len(input.Items))
	for i, item := range input.Items {
		if item.ProductID == "" {
			return nil, fmt.Errorf("item[%d].productId is required", i)
		}
		if item.Quantity <= 0 {
			return nil, fmt.Errorf("item[%d].quantity must be greater than 0", i)
		}
		if item.Price < 0 {
			return nil, fmt.Errorf("item[%d].price must be >= 0", i)
		}
		items[i] = domain.OrderItem{
			ProductID: item.ProductID,
			Quantity:  item.Quantity,
			Price:     item.Price,
		}
	}

	order := domain.NewOrder(input.CustomerID, items)

	if err := uc.repo.Save(ctx, order); err != nil {
		return nil, fmt.Errorf("failed to save order: %w", err)
	}

	return &CreateOrderOutput{
		OrderID:     order.OrderID,
		CustomerID:  order.CustomerID,
		TotalAmount: order.TotalAmount,
		ItemsCount:  order.ItemsCount,
		CreatedAt:   order.CreatedAt.Format("2006-01-02T15:04:05Z"),
	}, nil
}
